### PR TITLE
Add secretType

### PR DESCRIPTION
**Pull request by riverzhang**
* Add secretType in secret list

### CODE REVIEW

**Comment by cnwalker** [comment]
@riverzhang It seems like your image didn't upload

**Comment by riverzhang** [comment]
screenshot from 2017-08-11 19-26-35

**Comment by riverzhang** [comment]
@cnwalker Thanks

**Comment by riverzhang** [comment]
@maciaszczykm

**Comment by maciaszczykm** [comment]
@riverzhang Same things to fix as in your second pull request:

1. Missing translations.
2. Is this field visible also in details view?
3. Did you test on `serve:prod`?

**Comment by riverzhang** [comment]
@maciaszczykm     "translated is missing"     I do not understand what it means

File: src/app/frontend/secret/list/cardlist.html [hunk 1]
```diff
@@ -43,7 +43,13 @@
       [[Namespace|Label 'Namespace' which appears as a column label in the table of secrets (secret list view).]]
     </kd-resource-card-header-column>
     <kd-resource-card-header-column size="small"
-                                    grow="1"
+                                    grow="2"
+                                    sortable="true"
+                                    sort-id="$root.SortableProperties.SecretType">
```
Comment (maciaszczykm) [inline src/app/frontend/secret/list/cardlist.html]: It won't work. Remove it.

**Comment by maciaszczykm** [comment]
Why did you remove namespace cloumn? Sorting won't work, not namespace.

File: src/app/frontend/secret/list/cardlist.html (hunk 1 above)
Comment (riverzhang) [inline src/app/frontend/secret/list/cardlist.html]: @maciaszczykm   I am understand, I delete it, now it is ok.

File: src/app/frontend/secret/list/cardlist.html [hunk 2]
```diff
@@ -32,18 +32,18 @@
...
-                                    grow="4"
+                                    grow="2"
                                     sortable="true"
                                     sort-id="$root.SortableProperties.NAME">
       [[Name|Label 'Name' which appears as a column label in the table of secrets (secret list view).]]
     </kd-resource-card-header-column>
     <kd-resource-card-header-column size="small"
                                     grow="2"
-                                    ng-if="$ctrl.areMultipleNamespacesSelected()">
-      [[Namespace|Label 'Namespace' which appears as a column label in the table of secrets (secret list view).]]
```
Comment (maciaszczykm) [inline src/app/frontend/secret/list/cardlist.html]: Why do you remove namespace?

Comment (riverzhang) [inline src/app/frontend/secret/list/cardlist.html]: I dont know.  I run the dashboard and test to switch the namespace, it seems it is needed

File: src/app/frontend/secret/list/cardlist.html [hunk 3]
```diff
@@ -32,7 +32,7 @@
   </kd-zero-state>
   <kd-resource-card-header-columns ng-show="$ctrl.secretList.listMeta.totalItems">
     <kd-resource-card-header-column size="small"
-                                    grow="4"
+                                    grow="2"
```
Comment (maciaszczykm) [inline src/app/frontend/secret/list/cardlist.html]: Maybe 3?

File: src/app/frontend/secret/list/cardlist.html [hunk 4]
```diff
@@ -43,7 +43,12 @@
       [[Namespace|Label 'Namespace' which appears as a column label in the table of secrets (secret list view).]]
     </kd-resource-card-header-column>
     <kd-resource-card-header-column size="small"
-                                    grow="1"
+                                    grow="2"
+                                    sort-id="$root.SortableProperties.SecretType">
```
Comment (maciaszczykm) [inline src/app/frontend/secret/list/cardlist.html]: `sort-id` is not needed here.

Comment (riverzhang) [inline src/app/frontend/secret/list/cardlist.html]: ok   I delete it

File: src/app/frontend/secret/list/cardlist.html (hunk 3 above)
Comment (riverzhang) [inline src/app/frontend/secret/list/cardlist.html]: 3 and 2 are the distance between the two？

Comment (riverzhang) [inline src/app/frontend/secret/list/cardlist.html]: I run  daborad and  test it,  grow="2"  is better.
screenshot from 2017-08-16 21-02-46

File: src/app/backend/resource/secret/list.go [hunk 5]
```diff
@@ -65,8 +65,9 @@ func (spec *ImagePullSecretSpec) GetData() map[string][]byte {
 
 // Secret is a single secret returned to the frontend.
 type Secret struct {
-	api.ObjectMeta `json:"objectMeta"`
-	api.TypeMeta   `json:"typeMeta"`
+	ObjectMeta    api.ObjectMeta  `json:"objectMeta"`
+	TypeMeta      api.TypeMeta    `json:"typeMeta"`
```
Comment (maciaszczykm) [inline src/app/backend/resource/secret/list.go]: Why did you change it?

File: src/app/backend/resource/secret/list.go [hunk 6]
```diff
@@ -65,8 +65,9 @@ func (spec *ImagePullSecretSpec) GetData() map[string][]byte {
...
+	SecretType    v1.SecretType   `json:"secretType"`
```
Comment (maciaszczykm) [inline src/app/backend/resource/secret/list.go]: Use `string` as a type?

File: src/app/backend/resource/secret/list.go (hunk 5 above)
Comment (riverzhang) [inline src/app/backend/resource/secret/list.go]: Need to  the same definition as the secrettype

File: src/app/backend/resource/secret/list.go (hunk 6 above)
Comment (riverzhang) [inline src/app/backend/resource/secret/list.go]: yes

Comment (riverzhang) [inline src/app/backend/resource/secret/list.go]: clientgo   define      type SecretType string

Comment (maciaszczykm) [inline src/app/backend/resource/secret/list.go]: So use `string` on our side.

Comment (riverzhang) [inline src/app/backend/resource/secret/list.go]: ok

Comment (riverzhang) [inline src/app/backend/resource/secret/list.go]: @maciaszczykm

Comment (maciaszczykm) [inline src/app/backend/resource/secret/list.go]: You did not resolve my other comments...

File: src/app/externs/backendapi.js [hunk 7]
```diff
@@ -1173,6 +1174,7 @@ backendApi.Secret;
  * @typedef {{
  *   secrets: !Array<!backendApi.Secret>,
  *   listMeta: !backendApi.ListMeta,
+ *   secretType: string,
```
Comment (maciaszczykm) [inline src/app/externs/backendapi.js]: Can you show me where do you have this in secret list API endpoint?

File: src/app/frontend/secret/detail/info.html [hunk 8]
```diff
@@ -19,5 +19,10 @@
   <kd-info-card-section>
     <kd-object-meta-info-card object-meta="::$ctrl.secret.objectMeta">
     </kd-object-meta-info-card>
+    <kd-info-card-entry title="[[Type|Secret info details section type entry.]]">
+      <kd-middle-ellipsis display-string="{{::$ctrl.secret.type}}">
+      </kd-middle-ellipsis>
+      <div ng-hide="::$ctrl.secret.type">-</div>
```
Comment (maciaszczykm) [inline src/app/frontend/secret/detail/info.html]: This does not match https://github.com/kubernetes/dashboard/pull/2256/files#diff-a2f3442e2ae2857ec87f5b2601a01f96R1187.

```

File: src/app/backend/resource/secret/list.go (hunk 6 above)
Comment (maciaszczykm) [inline src/app/backend/resource/secret/list.go]: You can use string. There is no restriction to `v1.SecretType`.

File: src/app/externs/backendapi.js [hunk 9]
```diff
@@ -1182,6 +1184,7 @@ backendApi.SecretList;
  * @typedef {{
  *   objectMeta: !backendApi.ObjectMeta,
  *   typeMeta: !backendApi.TypeMeta,
+ *   secretType: string
```
Comment (maciaszczykm) [inline src/app/externs/backendapi.js]: Why?

**Comment by maciaszczykm** [comment]
Test on `serve:prod` before pinging us.

File: src/app/frontend/secret/detail/info.html (hunk 8 above)
Comment (riverzhang) [inline src/app/frontend/secret/detail/info.html]: type Secret struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
	Data map[string][]byte `json:"data,omitempty" protobuf:"bytes,2,rep,name=data"`
	StringData map[string]string `json:"stringData,omitempty" protobuf:"bytes,4,rep,name=stringData"`
	Type SecretType `json:"type,omitempty" protobuf:"bytes,3,opt,name=type,casttype=SecretType"`
}

const MaxSecretSize = 1 * 1024 * 1024
type SecretType string

====
use type  not SecretType?

File: src/app/backend/resource/secret/list.go (hunk 6 above)
Comment (riverzhang) [inline src/app/backend/resource/secret/list.go]: it is define use type not to const,  so it is  error when I build code.

File: src/app/externs/backendapi.js (hunk 9 above)
Comment (riverzhang) [inline src/app/externs/backendapi.js]: // Secret holds secret data of a certain type. The total bytes of the values in
// the Data field must be less than MaxSecretSize bytes.
type Secret struct {
	metav1.TypeMeta `json:",inline"`
	// Standard object's metadata.
	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
	// +optional
	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`

	// Data contains the secret data. Each key must consist of alphanumeric
	// characters, '-', '_' or '.'. The serialized form of the secret data is a
	// base64 encoded string, representing the arbitrary (possibly non-string)
	// data value here. Described in https://tools.ietf.org/html/rfc4648#section-4
	// +optional
	Data map[string][]byte `json:"data,omitempty" protobuf:"bytes,2,rep,name=data"`

	// stringData allows specifying non-binary secret data in string form.
	// It is provided as a write-only convenience method.
	// All keys and values are merged into the data field on write, overwriting any existing values.
	// It is never output when reading from the API.
	// +k8s:conversion-gen=false
	// +optional
	StringData map[string]string `json:"stringData,omitempty" protobuf:"bytes,4,rep,name=stringData"`

	// Used to facilitate programmatic handling of secret data.
	// +optional
	Type SecretType `json:"type,omitempty" protobuf:"bytes,3,opt,name=type,casttype=SecretType"`
}

const MaxSecretSize = 1 * 1024 * 1024

type SecretType string

=========
 secretType: !backendApi.SecretType?

File: src/app/frontend/secret/detail/info.html (hunk 8 above)
Comment (maciaszczykm) [inline src/app/frontend/secret/detail/info.html]: I do not understand your point at all. You have posted Kubernetes Secret definition, not one from our backend.

File: src/app/externs/backendapi.js (hunk 9 above)
Comment (maciaszczykm) [inline src/app/externs/backendapi.js]: What is this? You modified `IngressDetail`, so is not related at all. It looks like you are pasting random code snippets right now...

File: src/app/externs/backendapi.js (hunk 7 above)
Comment (maciaszczykm) [inline src/app/externs/backendapi.js]: @riverzhang

Comment (riverzhang) [inline src/app/externs/backendapi.js]: ok!  This is not needed here, I delete it.

File: src/app/frontend/secret/detail/info.html (hunk 8 above)
Comment (riverzhang) [inline src/app/frontend/secret/detail/info.html]: ok

File: src/app/externs/backendapi.js (hunk 9 above)
Comment (riverzhang) [inline src/app/externs/backendapi.js]: ok

File: src/app/frontend/secret/list/card.html [hunk 10]
```diff
@@ -32,6 +32,12 @@
         </kd-middle-ellipsis>
       </div>
     </kd-resource-card-column>
+    <kd-resource-card-column ng-if="$ctrl.secret.type">
```
Comment (maciaszczykm) [inline src/app/frontend/secret/list/card.html]: This can hide the whole column and destroy table view. Column header has no "if" and if this will be hidden, the table will be corrupted.

**Comment by maciaszczykm** [comment]
I can also see, that you have provided translations for both Chinese and Japanese. Do you actually know both languages?

**Comment by riverzhang** [comment]
@maciaszczykm I understand Chinese, do not understand Japanese, there is a small amount of Japanese, I will ask my friend to translate it

File: src/app/frontend/secret/list/card.html (hunk 10 above)
Comment (riverzhang) [inline src/app/frontend/secret/list/card.html]: ok! like  this
 <kd-resource-card-column>
      <kd-middle-ellipsis display-string="{{$ctrl.secret.type}}">
      </kd-middle-ellipsis>
      <div ng-hide="::$ctrl.secret.type">-</div>
    </kd-resource-card-column>
